### ISSUE TITLE
[server plugins] Add greater control over FFT windowing

### ISSUE THREAD

**Issue by brianlheim** [supercollider/supercollider#4130]
See https://www.listarc.bham.ac.uk/lists/sc-dev/msg58648.html

Expected Behavior
-----------------

Add support for greater control over the window types used in the FFT / IFFT plugins. Specifically, add [Tukey window](https://en.wikipedia.org/wiki/Window_function#Tukey_window). Perhaps I'm misunderstanding but I think we already offer pre/post windowing by allowing FFT and IFFT to be separate. @joslloand may be able to correct me there. 

Current Behavior
----------------

Can only use rectangular, sine, and Hann windows.

cc @jamshark70 @joslloand

**Comment by joslloand** [comment]
@jamshark70, yep, this is correct. The Tukey "pre-window" is followed by the zero padded Hann "post-window".

To offer a most general functionality, separate pre and post windows would be specified, and/or passed in as arguments.

**Comment by Sciss** [comment]
Correct me if I'm wrong, but wouldn't it suffice to have a separate window-multiplication UGen, like

    MulWindow.ar(in, winSize, stepSize, type, parameter)

You will anyway need an additional input for window-specific parameter in the case of tukey and others (e.g. kaiser)

**Comment by brianlheim** [comment]
> To offer a most general functionality, separate pre and post windows would be specified, and/or passed in as arguments.

Just want to be absolutely clear-- you want a preWindow and a postWindow argument both in the FFT UGen? I do not know much about this domain, excuse my ignorance.

**Comment by brianlheim** [comment]
> You will anyway need an additional input for window-specific parameter in the case of tukey and others (e.g. kaiser)

Should the parameter always be a real number? Any possibility a window will need 2 parameters?

Other questions @joslloand @Sciss:
- is this headed in the direction of letting users supply a buffer directly for windowing?
- if so should/how could we compromise between the two extremes? (preprogrammed windows vs total freedom)

**Comment by Sciss** [comment]
Theoretically there are window functions with multiple parameters (https://en.wikipedia.org/wiki/Window_function#A_list_of_window_functions), but in practice I think there is usually only zero or one arguments. Note that for reconstruction you have the `IFFT` UGen, so you already have a separation of analysis and synthesis windows.

Of course, if you had a user-supplied buffer (i.e. including the assumption that it is being written and read by UGens other than `FFT`, `IFFT`), then you could have a case of two UGens for forward and two UGens for backwards transform, two that do the overlapping copying / adding and windowing, and two that do the transforms. Not sure if many people would use that, though, or if adding just more window types and a window parameter to `FFT`, `IFFT` is the better / simpler solution.
